rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+# RFC 9292 - 2021-09-21 - `throttle` transform
+
+This RFC proposes the addition of a new transform that provides a user the ability to control the throughput of specific event streams.
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
+
+## Pain
+
+- Users lack the necessary tooling to control throughput which can cause excessive costs and negatively impact downstream services due to increased load
+- Admins cannot set quotas for users/usergroups utilizing Vector
+
+## Proposal
+
+### User Experience
```

Comment:
From reading this, it's not entirely clear to me what happens when events are rate-limited.

Do we create back pressure (it appears not, given that it's in the "out of scope" section)? Do we always drop throttled events straight away, or do we re-apply them later? If so, is there an upper limit to the number of events we keep in the queue? Can users opt to re-route throttled events (should we list this as a potential future improvement?).

Also, is there any tie-in with the proposed "dead-letter pipeline"? Should we add it in the context session to have a reference back to that feature?

Reply:
As is, events over the rate limit are immediately dropped - noting "dead-letter pipeline" in the context makes sense though

Reply:
Updated in 3b135aefc